### ISSUE TITLE
Renaming a tiddler break links, the user should be warned at least

### ISSUE THREAD

**Issue by mildred** [Jermolene/TiddlyWiki5#196]
Renaming a tiddler break links. However, no hint about that is shown to the user when the title is modified.

Possible solutions in increasing order of complexity:
- explain what happens if the tiddler name is modified (links will be broken), discourage renaming tiddlers.
- display the list of backlinks that will be broken
- modify the linking tiddlers so the links are not brokens, take care of sub-tiddlers as well if there is any.

**Comment by edrex** [comment]
@Jermolene, [$:/core/modules/widgets/navigator.js](https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/widgets/navigator.js) handles `tm-rename-tiddler` events by calling [`$tw.wiki.renameTiddler`](https://github.com/Jermolene/TiddlyWiki5/blob/72941f3e9a916948222ff7be540d896e2b1901f6/core/modules/wiki-bulkops.js#L18) which updates tags and other references immediately, but it's not triggered when the title is changed in tiddler edit mode. Is that planned? 

I ask because I want to port [mGSD](https://github.com/simonbaird/mGSD)'s [RenameTagsPlugin](https://github.com/simonbaird/MPTW/blob/master/core/RenameTagsPlugin.js) UX (trigger reference updates on any rename, after prompting user) [to gsd5](https://github.com/roma0104/gsd5/issues/76) [1], and if that were in core it would at least give me a pattern to follow.

The major difficulty is getting code to run when a rename happens. I can see a few possibilities:

* `$tw.hooks.addHook("th-saving-tiddler",function(tiddler) {...` This is what I tried first, but the old title isn't available here.
* Replace `$tw.Wiki.addTiddler` with a wrapped version that detects renames and fires an event, I guess via a `wikimethod` module.
* Do something at the widget level with the title field on the edit form. I'm guessing this is more idiomatic TW5, although I'm not sure (and I don't know how to go about it).

IDK TW5 at all, just fumbling around in the dark here. Any advice greatly appreciated. Generally it seems like there should be more tiddler lifecycle hooks, but probably I'm missing something.

[1] gsd5 references aren't tags like mGSD's but regular fields, so `$tw.wiki.renameTiddlers` wouldn't work verbatim. I would probably register another listener for `tm-rename-tiddler` to do the job.

cc @roma0104 @simonbaird

**Comment by roma0104** [comment]
Thanks for the feedback both here and on GSD5.
This is an issue that I have appear from time to time.  In fact just yesterday I had this issue with a project.  

I haven't delved deep to work on this issue but my initial thought is how I can capture the title of the Project prior to the hook, `th-saving-tiddler`, being called.  It appears that the [draft.of field is removed](https://github.com/Jermolene/TiddlyWiki5/commit/424b2fea3279c14daac0b0ca2baabc7c8ad7d41e#diff-94f2ff697f3cad8f798cbe4f8d77ee3dR348) prior to the [hook being called](https://github.com/Jermolene/TiddlyWiki5/commit/424b2fea3279c14daac0b0ca2baabc7c8ad7d41e#diff-94f2ff697f3cad8f798cbe4f8d77ee3dR350).  Not certain of the ramifications of passing these fields to any hooked plugins would be to the saving mechanism, then we could have cleanup mechanism that does a sanity check before saving it to the store.

**Comment by edrex** [comment]
I'm not seeing a draft.of field on tiddler in the debugger:

```
tiddler: $tw.Tiddler
    fields: Object
        created: Thu Jan 14 2016 08:15:38 GMT-0800 (PST)
        modified: Thu Jan 14 2016 08:15:42 GMT-0800 (PST)
        title: "FooRenamed"
```

**Comment by edrex** [comment]
In the interest of not hijacking this issue too entirely, I'd like to propose the following UX for core tiddlywiki renames, identical to what's provided by `RenameTiddlerPlugin.js`:

- When the user confirms an edit to the title field in the edit form, gather a list of references as in `$tw.wiki.renameTiddler`
- Present a summary to the user in a confirmation dialog:

```
Would you like to update FooTiddler references to BarTiddler in 6 other tiddlers?
   - BazTiddler
   - BanTiddler
   ...
```
 - If user confirms, make the update. Otherwise don't, but still save the current tiddler I guess.

I think this is close to what @mildred was asking for in the original issue

**Comment by edrex** [comment]
I'd like to propose the following to address dangling tag and list references when renaming a tiddler. It's the spec for #2223. Please provide feedback.

When a user modifies the title of an existing tiddler on the edit form, after checking for rename conflict and if so confirming overwrite,

- [ ] Fire a hook `th-renaming-tiddler`, modeled after `th-saving-tiddler`, but with `from_title` set to the old name.
- [x] Search the store for any tiddlers with tags or lists referencing the old name for the current tiddler

Note that if there are existing tag/list references to the new name (regardless of whether an existing tiddler was overwritten), these will remain. This is the current behavior and is considered "by design".

What do people think? I am hoping we can get by here without a prompt, but am not taking a hard line on it. Are there real examples where a prompt would be needed?

**Comment by felixhayashi** [comment]
> Fire a hook th-renaming-tiddler, modeled after th-saving-tiddler, but with from_title set to the old name.

:+1: 

> Search the store for any tiddlers with tags or lists referencing the old name for the current tiddler

…and replace them.

> Note that if there are existing tag/list references to the new name (regardless of whether an existing tiddler was overwritten), these will remain.

:+1: I agree!

---

**In addition:** As said in https://github.com/Jermolene/TiddlyWiki5/issues/1614#issuecomment-171933937, and in corrspondence to the OP, I would also highly welcome if not only the tags and list fields were updated but a redirect of links (using a regex-based replacement).*

**Comment by edrex** [comment]
Thanks for the feedback, it hadn't occurred to me that wikitext links need to be updated as well. How about we do that as a follow-on patch?

**Comment by felixhayashi** [comment]
Hi @edrex 

> How about we do that as a follow-on patch, using the hook?

that would be nice.

Since `retargetReferences` is already called when a renaming event occurs, I'd highly favor to implement a link rename operation directly in the function `retargetReferences`, which you introduced in your existing patch.

**Comment by tobibeer** [comment]
I think it needs to be a modal, especially so that there is an awareness of all the tidbits that are affected. If a user wants to, they can configure it to do its thIng silently. I would not want that.

**Comment by edrex** [comment]
@tobibeer, I'm interested in specific cases where automatically updating references might be harmful. Can you think of any? If we can update references safely, then it seems there is no need to disturb the end user's flow with a modal.

**Comment by felixhayashi** [comment]
In my opinion a rename operation should never invoke any modal. At most a notify() but that is already too much.

> I think a modal is only needed if we can't update references safely, which is why I'm interested in specific cases where automatic updates might be harmful. 

@edrex I would not say a rename can be harmful, even if it only replaces a subset of links, it is better than nothing. The real challenge is to consider the million ways a user might reference / link to a tiddler from another tiddler. No way a rename can detect all of these and even if it could, what should it do about it? I think this is the reason why @Jermolene hasn't picked up this issue yet, because it is not really clear what to rename and how. A simple regex search an replace over tiddler fields doesn't work :(

It is ease to replace title here:

* `<$link to="NAME" …`
* `[[NAME]]`

but what about  `<$vars foo="NAME" <$link to=<<foo>> …`?

**Comment by tobibeer** [comment]
The function of a modal / wizard would NOT be to inform a user as to what has been automagically replaced … but to allow a user to (de)select the ones he or she would (not) want modified.

**Comment by pmario** [comment]
-1 for a modal. I'd like a summary tiddler similar to the import tiddler, where I can select content and even have a look. 

The import mechanism seems to be good here. I can manually change the content on a per tiddler basis. when I'm done I can impot it.

**Comment by tobibeer** [comment]
I'm fine with it not being a modal. That' snot the crucial bit. The crucial but would be that I get to review and decide and an awareness of what gets updated.

**Comment by edrex** [comment]
I've put up a smaller pull request #2247 with just the hook, so editions can start experimenting and as a step toward reference updating in core.

Rather than add a new renaming hook as in the previous pull request, this just adds a field `tiddler.renaming` to `th-saving-tiddler` in case of a rename. Review greatly appreciated!

@tobibeer and @felixhayashi, I appreciate both your perspectives and like that you have strong views about the UX. I think since Tiddlywiki is a multi-purpose environment, different users are inevitably going to want different workflows. The above change can support either so go review it! :)

Back on topic, I'd like to distinguish field updating from body text updating, since I see the first as lower risk (agree?). I think conflating the two is part of why this issue hasn't gotten much traction. I could see wanting automatic updates for atomic fields (or a single all-or-nothing prompt, with an option to skip the prompt) along with a wizard for find and replace updates, for example.

**Comment by felixhayashi** [comment]
Hi @edrex 

sorry for the criticism but I think the hook in #2223 is far more reasonable, especially because it is temporary, doesn't leave traces and is backwards compatible. Many people highly dislike to have logging data in their wiki about what was renamed into what. Also this solution is not backwards compatible since many people may already use a renaming field for other purposes :(

I appreciate your work very much, but I would suggest to stick with a hook, which is great.

> Back on topic, I'd like to distinguish atomic field updating from body text updating, since I see the first as lower risk (agree?).

Not quite sure if I understood you correctly, but I even favor a solution where the user doesn't even have to say yes/no but the task is just executed (in an atomic way) and only if the atomic transaction fails (unlikely, why should it fail?) the user is informed.

**Comment by edrex** [comment]
@felixhayashi, if the criticism is meant to move the issue forward, I welcome it so don't apologize.

If you'll read the latest patch closely you'll see that it sets a temporary property `tiddler.renaming`, not a property of `tiddler.fields`, so it won't conflict with a field named `renaming` and it won't persist. If you could provide further feedback on the PR issue that would be preferred.

**Comment by felixhayashi** [comment]
> If you'll read the latest patch closely you'll see that it sets a temporary property tiddler.renaming

Ah, so it gets added to the changedtiddlers refresh object, then its ok of course! reminds me of my own quest: #1550: "ChangedTiddlers: append previous state (tiddler object) to change object" ;)

**Comment by Jermolene** [comment]
In terms of the UI for handling tiddler renames, I'm not wildly keen on modals. The trouble is that it is hard to refer to the rest of the wiki to check things in order to be able to respond to the modal. I prefer something inline and modeless. For example, we could detect that the title has been changed in edit mode and display a special panel underneath the title that explains what's going on and provides checkboxes for whether to rename references as tags, lists, etc.

**Comment by felixhayashi** [comment]
I agree, I also would not want modals, not even reports/summaries of what has happened.

>  For example, we could detect that the title has been changed in edit mode and display a special panel underneath the title that explains what's going on and provides checkboxes for whether to rename references as tags, lists, etc.

A problem that I see with this method is that when a tiddler has been renamed from "outside". E.g. TiddlyMap can be used to rename tiddlers from within the map. In my opinion, TW should just do its best to rename references everywhere where it is obvious that things should be renamed. 

In addition, it should cleary state in the docs what is happening when a user renames a tiddler.

What would be great if the renameTiddler function could be enhanced by also replacing explicitly stated links and transcludes i.e. `[[ … ]]` and `{{…}}`.

**Comment by Jermolene** [comment]
> TiddlyMap can be used to rename tiddlers from within the map

Do multiple tiddlers get renamed in one operation, or just a single tiddler? If the logic I referred to above was easily transcludable, could you just include it in your dialogue?

> What would be great if the renameTiddler function could be enhanced by also replacing explicitly stated links and transcludes i.e. [[ … ]] and {{…}}.

There's some significant complexity to that. We could use a regexp but there would be false positives. Perhaps we could use the parse tree to find the links and transclusions, as we do for computing backlinks.

Anyhow, one way to treat these things is as a bulk operation that needs confirmation: that's the idea of the `$:/Import` tiddler: it's the beginnings of a general purpose way of displaying the differences between a set of incoming tiddlers and the current state of the store, and then accepting or rejecting the changes. My plan is to use the same thing for displaying the results of a find and replace operation.

**Comment by edrex** [comment]
I had been thinking about an API for responding to renames that would support all the various UX possibilities. 

It would be a hook `respondToRename(oldName, newName, targetTiddler)`, where `targetTiddler` is the tiddler to be considered for update. Implementers would return a sparse object with just the fields to be changed set. This would allow hook invoker to display a list of changes to the user and then apply them in a separate step. It would also allow an alternate invoker which would apply changes without prompting. I like the idea of keeping the API and the UX decoupled like this, since it increases the flexibility of the system.

**Comment by edrex** [comment]
I missed the last part of Jeremy's last comment, and I hadn't had a look at `$:Import`. It sounds like object diffing, and somewhat related to my idea for an API for proposing automatic updates.

**Comment by nergal-perm** [comment]
As I understand the relinking task is very complex because of variety of ways the link can be created. Well, perhaps we could start by enumerating different ways of creating links and then estimating / guessing, how many links are created using those ways. Then it wil be possible to start working on relinking task, starting by the most frequent way of creating links. 

However, relinking direct links (`[[link]]`, `[[title|link]]`, `[title](#link)` for Markdown) seems very straightforward, isn't it? So this kind of relinking could be included in TW in almost no time to improve user experience.

**Comment by BurningTreeC** [comment]
@Jermolene oh yes that'll be good

**Comment by flibbles** [comment]
I've got a plugin for this now which I'm peddling wherever I can. [The new Relink plugin](https://flibbles.github.io/tw5-relink/) should be an answer to this problem. It handles renaming tiddler title occurrences in widgets, transclusions, links, filters, and fields.

**Comment by felixhayashi** [comment]
Really great stuff @flibbles and also a very well designed demo. :+1: Thanks.

**Comment by saqimtiaz** [comment]
> I'd actually still like to offer relinking-on-rename when we've got search-and-replace up and running.

@Jermolene does this still match your current thinking or should this issue be closed?

**Comment by Jermolene** [comment]
> @Jermolene does this still match your current thinking or should this issue be closed?

Yes, I think so. I like the idea of being able to review relinking changes before committing them.

**Comment by pmario** [comment]
>> @Jermolene does this still match your current thinking or should this issue be closed?
>
>Yes, I think so. I like the idea of being able to review relinking changes before committing them.

1) I think the relink-plugin does show a detailed info where changes are applied, when a tiddler is renamed
2) It contains a detailed overview, where "backlinks" happen in the tiddler info area. 

I think this issue can be closed. 

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/c747b3f8-1377-4bd4-8790-fcb541d6b852)

**Comment by Jermolene** [comment]
Hi @pmario I think this ticket still stands because the functionality is not in the core.

**Comment by nathanKrupla** [comment]
I'm in the process of documenting and world building for a book idea I've been thinking about for years.  As such, I'm renaming things often and I'm a bit perplexed right now.  I renamed a character (and their Tiddler) and found out shortly afterward that I'd broken the link...  they don't auto-remap.

I know this isn't the most productive question but how is it possible that such an essential piece of functionality hasn't been implemented in spite of more than 10 years of discussion over it?  I mean... we've seen a multitude of very intelligent suggestions above and people are still having to go to third party plugins?  I get it, the people on here live to write scripts to do stuff for them, but some of us just want the thing to work out of the box and this is the kind of functionality I design into my own stuff from day one.

I'd offer to write something, but I'm completely new to this codebase and I know absolutely jack about web development.  Most of my experience is custom interfaces for PLCs and robotics.  (No point in talking to a browser, just more to maintain.

So is there a solution coming for this, or do I need to hack something together?  If it's the latter, I'm probably going to go looking for something else.  The idea of this program is fantastic (I'd be willing to pay for it, honestly) but something as massive as this is a deal-breaker.  I'm sure there's thousands who've abandoned it for this reason as well.

PS I created this account today as I realized I hadn't made one for my new permanent email.  That's my real name.  Feel free to reach out on other platforms as I generally ignore Github's emails.

**Comment by flibbles** [comment]
I feel you.

Six years ago, I thought the same thing you did, which was, "Jeeze, TiddlyWiki really needs relinking. Maybe I just hack something together myself."

This led me to over 300+ man-hours of work, which turned me into a TiddlyWiki power user with dozens of PRs merged and a lot of *strong opinions* about the direction TiddlyWiki's development should be going. It turns out, Relinking was a **super difficult problem**. But the result was the [Relink plugin](https://flibbles.github.io/tw5-relink/), which is the comprehensive relinker you're looking for.

I personally don't have a single project that doesn't use my own plugin, and I think the TW community as a whole considers Relink as an essential plugin. Many people before you have argued that it should be part of the core, and I've frequently considered whether I should champion this motion, but ultimately I've concluded that Relink works better as a plugin. The reason why is because Relink adds complexity which requires participation from end-users which newbies won't know to do.

To explain, let's take an example that runs with the world-building theme of what you're doing. Suppose we have this macro.

```
\procedure town(name, country, occupants, description)

!!! <<name>>

''Description:'' <<description>>

''Country'': <$link to=<<country>> />

''Population:'' {{{ [subfilter<occupants>] }}}
\end

<<town
	Ockorock
	Toril
	"[tag[Orc]] [tag[Ogre]]"
	"""This town is home to many brutish races. It was founded after the [[Goblin Wars]]""" >>
```

Comprehensive relinking means Tiddlywiki would relink the tiddlers that appear in that procedure call, but how can it do that? Each parameter is different.

1. The `country` parameter corresponds to another single tiddler.
2. The `occupants` is a filter. Tiddlers used inside it are Orc and Ogre.
3. `description` is wikitext with a link referencing the tiddler "Goblin Wars".

Each parameter needs to be considered differently, and there's no way for Relinking to do that unless the user does something to describe the parameters. In Relink, the solutions is a `\relink` pragma.

```
\relink town country:title occupants:filter description:wikitext
\procedure town(name, country, occupants, description)
```

*Now* relinking can occur, but only if the end-user remains vigilant about maintaining these sorts of pragma. Without it, relinking is incomplete and unreliable. That is the way relinking is adding **complexity**.

Imagine how that works out for someone who's just getting into Tiddlywiki. They're learning how to use wikitext, and they research procedures and macros, and maybe there are notes in the documentation saying how "they need to be careful to include relink directives" along with an explanation why. And it's this extra complexity that might be immediately clear why it's important. Or they don't see that documentation at all, and they just mistakenly believe TiddlyWiki is a comprehensive relinker automatically. Neither of these situations are ideal.

Now consider it's a plugin.

1. New user uses Tiddlywiki for a bit, they realize that it needs to do relinking
2. They explore Tiddlywiki plugins and find one of the top rated plugins, Relink, does this.
3. They install it and adopt the behavior that the plugin requires.

Maybe I'm wrong, but I think that works better. The act of installing a plugin comes with an implicit, "*you must take some ownership over this*" which is required for comprehensive relinking.

What I think is important, and where TiddlyWiki may be dropping the ball a little, is that it needs to be *very easy* for new users to find/peruse/install plugins.

Anyway, those are my thoughts. Installing Relink should solve your problem. It's a simple drag and drop (unless you're using NodeJS, but then the Relink demo page has instructions for that too).
